eCommerce platforms/Adobe Magento/src/code/local/Microsoft/Dfp/Model/Observer.php

```diff
@@ -0,0 +1,251 @@
+ * It is also available through the world-wide-web at this URL:
+ * http://opensource.org/licenses/osl-3.0.php
+ *
+ * @category   Microsoft
+ * @package    Dynamics 365 Fraud Protection
+ * @copyright  Copyright (c) Microsoft Corporation. (https://www.microsoft.com)
+ * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+class Microsoft_Dfp_Model_Observer
+{
+	private $dfp;
+
+	function __construct()
+	{
+		$this->dfp = Mage::helper('dfp');
+	}
+
+	public function customerEventAfterAddressSave(Varien_Event_Observer $observer)
+	{
+		try {
+			if (Mage::app()->getRequest()->getRequestedRouteName() == 'customer') {
+
+				$this->dfp->log('************** customerEventAfterAddressSave ************* ');
+
+				if (empty(Mage::getSingleton('core/session')->getFptDfpSessionId())) {
+					return;
+				}
+				$customer = Mage::getModel('customer/customer')->load($observer->getCustomerAddress()->getCustomerId());
+
+				$this->dfp->log('************** Invoking UpdateAccount Address Change API - START ************* ');
+				$this->invokeUpdateAccountDFP($customer);
+				$this->dfp->log('************** Invoking UpdateAccount Address Change API - END ************* ');
+			}
+		} catch (exception $e) {
+			Mage::logException($e);
+		}
+	}
+
+	public function salesEventOrderAfterSave(Varien_Event_Observer $observer)
+	{
+		try {
+			$this->UpdateStatusToDFP($observer);
+		} catch (exception $e) {
+			Mage::logException($e);
+		}
+	}
+
+	private function UpdateStatusToDFP($observer)
+	{
+		$this->dfp->log("  -----------  UpdateStatusToDFP Observer Event START --------------  ");
+		$order = $observer->getEvent()->getOrder();
+		if ($order) {
+			if (!empty($order->getData()['status'])) {
+				$status = $order->getData()['status'];
+				if (!empty($status)) {
+					$this->dfp->log(" UpdateStatusToDFP Observer Event Status : " . $status);
+					switch (strtolower($status)) {
+						case 'order_approved':
+						case 'canceled':
+							$this->dfp->log("  -----------   Invoking PurchaseStatus API - START --------------  ");
+							$this->invokePurchaseStatusDFP($order);
+							$this->dfp->log("  -----------   Invoking PurchaseStatus API - END --------------  ");
+							break;
+						case 'refunded':
+							$this->dfp->log("  -----------   Invoking Refund API - START --------------  ");
+							$this->invokeRefundDFP($order);
+							$this->dfp->log("  -----------   Invoking Refund API - END --------------  ");
+							break;
+						case 'chargeback':
+							$this->dfp->log("  -----------   Invoking Chargeback API - START --------------  ");
+							$this->invokeChargebackDFP($order);
+							$this->dfp->log("  -----------   Invoking Chargeback API - END --------------  ");
+							break;
+					}
+				}
+			}
+		}
+
+		$this->dfp->log("  -----------  UpdateStatusToDFP Observer Event END --------------  ");
+	}
+
+	private function invokeRefundDFP($refund)
```

Comment:
Doesn't matter too much but is there any reason to not do some sort of singular invokeDFP with a payload and an event type?

Reply:
It's more akin to what someone would get if they were to auto-generate swagger classes.  Plus it also makes it easier to customize one event type or another, if needed.